lms/djangoapps/affiliates/models.py

```diff
@@ -0,0 +1,156 @@
+    website = models.CharField(null=True, blank=True, max_length=255, default='')
+    facebook = models.CharField(null=True, blank=True, max_length=255, default='')
+    twitter = models.CharField(null=True, blank=True, max_length=255, default='')
+    linkedin = models.CharField(null=True, blank=True, max_length=255, default='')
+    state = models.CharField(null=True, blank=True, default='na', choices=STATE_CHOICES, max_length=255)
+    country = CountryField(blank=True, null=True)
+
+    image = models.ImageField(upload_to=user_directory_path, null=True, blank=True)
+
+    members = models.ManyToManyField(User, through='AffiliateMembership')
+
+    def save(self, *args, **kwargs):
+        slug = slugify(self.name)
+
+        if AffiliateEntity.objects.filter(slug=slug).exclude(pk=self.pk).exists():
+            self.slug = '-'.join([slug, get_random_string(4)])
+        else:
+            self.slug = slug
+
+        super(AffiliateEntity, self).save(*args, **kwargs)
+
+
+    def delete(self):
+        with transaction.atomic():
+            self.courses.delete()
+            super(AffiliateEntity, self).delete()
+
+    class Meta:
+        unique_together = ('email', 'name')
+
+    def __unicode__(self):
+        return self.name
+
+    @property
+    def image_url(self):
+        if self.image:
+            return self.image.url
+        else:
+            return 'https://s3.amazonaws.com/fasttrac-beta/default_full.png'
+
+    @property
+    def memberships(self):
+        return AffiliateMembership.objects.filter(affiliate=self)
+
+    @property
+    def courses(self):
+        return CustomCourseForEdX.objects.filter(coach__in=self.members.all())
+
+
+
+class AffiliateMembership(models.Model):
+    role_choices = (
+        ('ccx_coach', 'Facilitator'),
+        ('instructor', 'Course Manager'),
+        ('staff', 'Program Director'),
+    )
+
+    member = models.ForeignKey(User)
+    affiliate = models.ForeignKey(AffiliateEntity, on_delete=models.CASCADE)
+    role = models.CharField(choices=role_choices, max_length=255)
+
+    @classmethod
+    def find_by_user(self, user):
+        return self.objects.get(member=user)
+
+
+
+@receiver(post_save, sender=AffiliateMembership, dispatch_uid="add_affiliate_course_enrollments")
+def add_affiliate_course_enrollments(sender, instance, **kwargs):
+    'Allow staff or instructor access to affiliate member into all affiliate courses if they are staff or instructor member.'
+    if not instance.role == 'ccx_coach':
+        for ccx in instance.affiliate.courses:
+            ccx_locator = CCXLocator.from_course_locator(ccx.course_id, ccx.id)
+            course = get_course_by_id(ccx_locator)
+
+            try:
+                with transaction.atomic():
+                    allow_access(course, instance.member, instance.role, False)
+            except IntegrityError:
+                print 'IntegrityError: Allow access failed.'
```

Comment:
Djangooooo... Sometimes works, sometimes throws this silly error. I handle it silently so it doesn't roll back the whole transaction.